src/lv_core/lv_refr.c

```diff
@@ -741,12 +741,80 @@ static void lv_refr_obj(lv_obj_t * obj, const lv_area_t * mask_ori_p)
     }
 }
 
+/**
+ * Rotate the VDB to the display's native orientation.
+ */
+static lv_color_t *lv_refr_vdb_rotate_pre(lv_disp_drv_t *drv, lv_area_t *area, lv_color_t *color_p) {
+    if(lv_disp_is_true_double_buf(disp_refr) && drv->sw_rotate) {
+        LV_LOG_ERROR("lv_refr_vdb_rotate_pre: cannot rotate a true double-buffered display!");
+        return color_p;
+    }
+    lv_coord_t area_w = lv_area_get_width(area);
+    lv_coord_t area_h = lv_area_get_height(area);
+    uint32_t total = area_w * area_h;
+    if(drv->rotated == LV_DISP_ROT_180) {
+        /* Swap the beginning and end values */
+        lv_color_t tmp;
+        uint32_t i = total - 1, j = 0;
+        while(i > j) {
+            tmp = color_p[i];
+            color_p[i] = color_p[j];
+            color_p[j] = tmp;
+            i--;
+            j++;
+        }
+        lv_coord_t tmp_coord;
+        tmp_coord = area->y2;
+        area->y2 = drv->ver_res - area->y1 - 1;
+        area->y1 = drv->ver_res - tmp_coord - 1;
+        tmp_coord = area->x2;
+        area->x2 = drv->hor_res - area->x1 - 1;
+        area->x1 = drv->hor_res - tmp_coord - 1;
+    } else if(drv->rotated == LV_DISP_ROT_90 || drv->rotated == LV_DISP_ROT_270) {
+        lv_color_t * orig_color_p = color_p;
+        color_p = _lv_mem_buf_get(sizeof(lv_color_t) * total);
```

Comment:
It seems it could request screen-sized buffers too. E.g. 480x320x2 = 300kB.

What about something like [this](https://forum.lvgl.io/t/setting-rotated-option-in-lv-disp-drv-t-causes-hard-fault/440/10) but writing the column into a buffer and passing the buffer to the flush_cb?

Reply:
I see what you mean. The memory usage is definitely a problem.

However, wouldn't your solution be too slow for practical use? If I'm understanding correctly, that would send one column at a time to `flush_cb`, which would cripple things like DMA-over-SPI transactions. Maybe I misunderstood.

--------------------------------------

There is an alternative method I discovered for implementing portrait rotations *in situ* (without the need for a second buffer). While rotating a rectangular buffer 90 degrees is very complicated to do in place, rotating a square buffer 90 degrees is quite trivial. To do this, LVGL needs to always send a square VDB to be flushed. Every rectangle would need to be divided into a number of squares. This method would use almost no additional memory but I think performance would be a problem.

Reference: first few paragraphs of https://en.wikipedia.org/wiki/In-place_matrix_transposition

Reply:
> If I'm understanding correctly, that would send one column at a time to flush_cb, which would cripple things like DMA-over-SPI transactions. Maybe I misunderstood.

SPI based display controller works like this:
1. you select a draw area (window)
2. send the image for the window in a bulk (the display controller will take care of beginning a new line when one is fully filled)

Therefore there is no difference in sending 1 row or 1 column. However, in stm32 MCU we use DMA to copy 1 line of VDB to the frame buffer. It's easy because it just copies adjacent bytes. Copying a column this way is quite problematic.

---

Square rotation sounds good. We can always crop a square which side length is equal to the shortest side of the rectangle. E.g.
1. Start with 100x80 rectangle
2. Use a 80x80 square, 20x100 rectangle remains  
3. 20x100x2 = 4,000, no too large, allocate a buffer and rotate there.

The worst case is when the buffer has `small x large` size because we can't crop large squares.  E.g. 20x800 (32k on 16 bit).
One crop lowers the size only with 800 bytes.

I can imagine an algorithm like this:
1. let's say the max rotate buffer is 10kB
2. if the VDB is smaller than 10 kB, rotate into the buffer
3. if the VDB is larger than 10 kB can we crop > 10kB area? 
   - if yes crop it and go back to the 2nd point
   - if we can't, rotate a 10 kB area into the buffer and go back to 2nd. point
  
What do you think?

Reply:
Hmm. I think that would be best implemented at the point where rendering gets split for small VDBs, otherwise I think there would need to be another linked list to track what parts of the VDB have been flushed thus far. I'm not as familiar with that part of the code so it will take some time for me to come up with a solution.

Reply:
What about defining a maximal rotation buffer (e.g. 10k) and splitting the original area into 10 k chunks. It seems to me this is the easiest to implement and configure too.

What do you think?

Reply:
So render into the VDB as normal, then when we get to `lv_refr_vdb_rotate_pre`, rotate it 10KB at a time? That sounds doable.

Reply:
Yes, rotate as many lines at once that fit into 10k.

Reply:
Alright; that was pretty straightforward. I've pushed the changes. Let me know if you think anything else needs tweaking.

90 and 270 degree rotation is still slower than I'd like, but since this is software rotation, I'm not sure we are going to get much else out of pure CPU computation. The user always has the option to implement a faster hardware-specific algorithm themselves by disabling `sw_rotate` and doing the rotation in their driver, which I think is an acceptable compromise.

Reply:
>  I'm not sure we are going to get much else out of pure CPU computation. The user always has the option to implement a faster hardware-specific algorithm themselves by disabling sw_rotate and doing the rotation in their driver, which I think is an acceptable compromise.

I definitely agree.

I'm just testing it.